mart/callbacks/anomalib.py

```diff
@@ -0,0 +1,182 @@
+    def on_after_batch_transfer(self, pl_module, batch, dataloader_idx):
+        batch = self._on_after_batch_transfer(batch, dataloader_idx)
+
+        # FIXME: Dispatch to on_after_train/val/test_batch_transfer
+        if not pl_module.trainer.testing:
+            return batch
+
+        # We clone these because they were created in inference mode and we
+        # in-place update them.
+        # FIXME: We probably need to undo any transform step on these tensors.
+        image, mask = batch["image"], batch["mask"]
+        image = image.clone()
+        mask = mask.clone()
+
+        device = image.device
+        bs, _, height, width = image.shape
+
+        assert height == width
+
+        # Create optimization variables for angle, hue, and saturation
+        angle = torch.tensor(
+            [self.angle_init] * bs,
+            device=device,
+            dtype=torch.float32,
+            requires_grad=True,
+        )
+
+        hue = torch.tensor(
+            [self.hue_init] * bs, device=device, dtype=torch.float32, requires_grad=True
+        )
+
+        sat = torch.tensor(
+            [self.sat_init] * bs, device=device, dtype=torch.float32, requires_grad=True
+        )
+
+        for step_idx in range(self.steps):
+            if step_idx % self.restart_every == 0:
+                optimizer = torch.optim.Adam(
+                    [
+                        {"params": angle, "lr": self.angle_lr},
+                        {"params": hue, "lr": self.hue_lr},
+                        {"params": sat, "lr": self.sat_lr},
+                    ],
+                    maximize=True,
+                )
+
+                # Randomly reinitialize parameters after restart
+                if step_idx > 0:
+                    angle.data.uniform_(*self.angle_bound)
+                    hue.data.uniform_(*self.hue_bound)
+                    sat.data.uniform_(*self.sat_bound)
+
+            # Clip parameters to valid bounds using straight-through estimator
+            clipped_angle = (
+                angle + (torch.clip(angle, *self.angle_bound) - angle).detach()
+            )
+            clipped_hue = hue + (torch.clip(hue, *self.hue_bound) - hue).detach()
+            clipped_sat = sat + (torch.clip(sat, *self.sat_bound) - sat).detach()
+
+            image_adv, mask_adv = self.attack(
+                image, mask, clipped_angle, clipped_hue, clipped_sat
+            )
+            # FIXME: Do we need to normalize image_adv?
+            pred_label, pred_mask = pl_module(image_adv)
+            loss = self.loss(pred_mask, mask_adv)
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+        batch["image"] = image_adv.detach()
+
+        return batch
+
+    def attack(self, image, mask, angle, hue, sat):
+        # Rotate image by angle degrees
+        theta = torch.deg2rad(angle)
+        image_rot = rotate_three_pass(image, theta, N=-1, padding_mode="replicate")
```

Comment:
This may have the in-place modification bug from `rotate_three_pass`, do you need `image.clone()` just to be safe?

Reply:
i think this is fixed in the latest code.